File: ml/udm/18_hierachical_clustering/hc.py
# Hierarchical Clustering

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:, [3, 4]].values

# using dentogram to find optimal number of clusters
import scipy.cluster.hierarchy as sch
dentogram = sch.dendrogram(sch.linkage(X, method='ward'))
plt.title('Dentogram')
plt.xlabel('Customers')
plt.ylabel('Euclidean distances')
plt.show()

# fitting model
from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters=5, affinity='euclidean', linkage='ward')
y_hc = hc.fit_predict(X)

# visualizing
plt.scatter(X[y_hc == 0, 0], X[y_hc == 0, 1], s=100, c='red', label='Cluster 1')
plt.scatter(X[y_hc == 1, 0], X[y_hc == 1, 1], s=100, c='blue', label='Cluster 2')
plt.scatter(X[y_hc == 2, 0], X[y_hc == 2, 1], s=100, c='green', label='Cluster 3')
plt.scatter(X[y_hc == 3, 0], X[y_hc == 3, 1], s=100, c='cyan', label='Cluster 4')
plt.scatter(X[y_hc == 4, 0], X[y_hc == 4, 1], s=100, c='magenta', label='Cluster 5')
plt.title('Cluster of clients')
plt.xlabel('Income (k$)')
plt.ylabel('Score (1-100)')
plt.legend()
plt.show()

File: ml/udm/18_hierachical_clustering/hc.R
# Hierarchical Clustering

# Importing the dataset
dataset = read.csv('Mall_Customers.csv')
X = dataset[4:5]

# using dentogram to find optimal cluster number
dentogram = hclust(dist(X, method = 'euclidean'), method = 'ward.D')
plot(dentogram)

# fitting model
y_hc = cutree(dentogram, k = 5)

# vsualizing 
library(cluster)
clusplot(
  X,
  y_hc,
  lines = 0,
  shade = TRUE,
  color = TRUE,
  labels = 2,
  plotchar = FALSE,
  span = TRUE,
  main = 'Cluster of client',
  xlab = 'Income',
  ylab = 'Score'
)